File: src/graph/prim/km53.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;
let vertexNum = 0;
let edgeNum = 0;
let edges: number[][] = [];

let minDist: number[] = [];
let inTree: boolean[] = [];

rl.on("line", (line) => {
  lineno++;

  if (lineno === 1) {
    [vertexNum, edgeNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));

    edges = Array.from({ length: vertexNum + 1 }, () =>
      Array.from<number>({ length: vertexNum + 1 }).fill(Infinity),
    );

    // minDist = Array.from({ length: vertexNum + 1 }, () => Infinity);
    minDist = Array.from({ length: vertexNum + 1 }, () => Infinity);
    minDist[1] = 0;

    inTree = Array.from<boolean>({ length: vertexNum + 1 }).fill(false);

    return;
  }

  const [left, right, len] = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));

  edges[left][right] = len;
  edges[right][left] = len;

  if (lineno === 1 + edgeNum) {
    rico();
    rl.close();
  }
});

function rico() {
  for (let i = 0; i < vertexNum - 1; i++) {
    let candidate = -1;
    let minVal = Infinity;

    for (let j = 1; j <= vertexNum; j++) {
      if (!inTree[j] && minDist[j] < minVal) {
        candidate = j;
        minVal = minDist[j];
      }
    }

    inTree[candidate] = true;

    for (let j = 1; j <= vertexNum; j++) {
      if (!inTree[j] && edges[j][candidate] < minDist[j]) {
        minDist[j] = edges[j][candidate];
      }
    }
  }

  let ans = 0;
  for (const item of minDist.slice(1)) {
    ans += item;
  }

  console.log(ans);
}

File: src/graph/prim/km53.py
from typing import List

# constants
Infinity = (1 << 31) - 1

# input
vertexNum, edgeNum = map(int, input().strip().split())

edges: List[List[int]] = [
    [Infinity for _ in range(vertexNum + 1)] for _ in range(vertexNum + 1)
]

for i in range(1, vertexNum + 1):
    edges[i][i] = 0

for _ in range(edgeNum):
    left, right, len = map(int, input().strip().split())
    edges[left][right] = len
    edges[right][left] = len

minDist: List[int] = [Infinity for _ in range(vertexNum + 1)]
inTree: List[bool] = [False for _ in range(vertexNum + 1)]

minDist[1] = 0
# inTree[1] = True

# for i in range(2, vertexNum + 1):
#     if edges[i][1] < minDist[i]:
#         minDist[i] = edges[i][1]


for _ in range(0, vertexNum - 1):
    candidate = -1
    minVal = Infinity

    for j in range(1, vertexNum + 1):
        if not inTree[j] and minDist[j] < minVal:
            candidate = j
            minVal = minDist[j]

    inTree[candidate] = True

    for j in range(1, vertexNum + 1):
        if not inTree[j] and edges[j][candidate] < minDist[j]:
            minDist[j] = edges[j][candidate]

ans: int = 0

for item in minDist[1:]:
    ans += item

print(ans)
